lib/pure/collections/smart_ptr.nim

```diff
@@ -0,0 +1,152 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2019 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+
+import allocators
+
+type
+  UniquePtr*[T] = object
+    ## non copyable pointer to object T, exclusive ownership of the object is assumed
+    val: ptr tuple[value: T, allocator: Allocator]
```

Comment:
I thought about this idea. IMO, it is bad if 2 smart pointers will be treated as incompatible types, just because allocator was different.
I see the following options:
1. No support for different allocators ;(
2. There is only one allocator returned by `getAllocator` and it user responsibility to set it to thread local allocator, global allocator or custom allocator at start up time.
3. Do what I just did. While overhead seems noticeable, it is possibly less than one might think.
There is 16 byte alignment requirement for all heap allocations on 64 bit platforms, hence you are likely to overallocate anyway hence there is a space for extra pointer.
4. Do what I just did, but do have mode were `Allocator` type is actually `void` and have zero byte size and alloc, dealloc are silently mapped to system's `alloc` and `dealloc`. Hence, those users who don't need customisation do not pay for it. Possible ABI compatibility problems due different defined flags used by different libraries. Problems are potentially solvable.

Reply:
I have some ideas of solving this problem... I'm writing a blog post.

Reply:
Cool, I will wait for it then. Or if it is possible to describe it in one sentence please do so

Reply:
Use a pool for the allocations, make the pool's address aligned so that you can bitmask the pointers to access the pool's header which contains the allocator.